### ISSUE TITLE
[BUG]: assert grad_chunk.l2_norm is not None

### ISSUE THREAD

**Issue by liangzz1991** [hpcaitech/ColossalAI#6102]
### Is there an existing issue for this bug?

- [X] I have searched the existing issues

### 🐛 Describe the bug

Modify the code to adapt to qwen2vl(transformers.Qwen2VLForConditionalGeneration) and find that the loss can be calculated, but partial chunk :  grad_chunk.l2_norm is None.........(LLM is ok)
![image](https://github.com/user-attachments/assets/b21cedda-30d3-4aad-9112-4c1ab0597211)

modified source code to print more information：
![image](https://github.com/user-attachments/assets/927e9fd6-7a14-4cac-a3cb-2bf7a098ceaa)
result:
![image](https://github.com/user-attachments/assets/c0517e1e-1ab1-498b-9d43-7229270126d1)

**Comment by Edenzzzz** [comment]
For this kind of issue, it makes more sense to share what modifications you made (since we don't maintain individual modifications).

**Comment by liangzz1991** [comment]
> modified s

I didn't modify the source code originally. The only place I modified was in the second picture, in order to print the error details.

**Comment by liangzz1991** [comment]
> ### Is there an existing issue for this bug?
> * [x]  I have searched the existing issues
> 
> ### 🐛 Describe the bug
> Modify the code to adapt to qwen2vl(transformers.Qwen2VLForConditionalGeneration) and find that the loss can be calculated, but partial chunk : grad_chunk.l2_norm is None.........(LLM is ok) ![image](https://private-user-images.githubusercontent.com/12033811/380106501-b21cedda-30d3-4aad-9112-4c1ab0597211.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5MzUyMjcsIm5iZiI6MTcyOTkzNDkyNywicGF0aCI6Ii8xMjAzMzgxMS8zODAxMDY1MDEtYjIxY2VkZGEtMzBkMy00YWFkLTkxMTItNGMxYWIwNTk3MjExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI2VDA5Mjg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmZDg2YmM5ODRmNzZhMzEwYzQ0OGMxOWI2NGU2YjEzYTE1NWIzOGQwZWRlNTgyMmRmOGYzYTcwN2RkYTllNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.POChlBWt-euU2dCatxMpnpdeIrPDOOQ9hYCxge5nGVc)
> 
> modified source code to print more information： ![image](https://private-user-images.githubusercontent.com/12033811/380107049-927e9fd6-7a14-4cac-a3cb-2bf7a098ceaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5MzUyMjcsIm5iZiI6MTcyOTkzNDkyNywicGF0aCI6Ii8xMjAzMzgxMS8zODAxMDcwNDktOTI3ZTlmZDYtN2ExNC00Y2FjLWEzY2ItMmJmN2EwOThjZWFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI2VDA5Mjg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ2YWZiMzFjZWRhMzVlMzA0MTY3NGQ2MWQ1ZTA0MWYxZGVhZjI5NjE1NmNhYjdkNDAxMDM0YmE2NjhmNDFlNDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.X3QT8yrE28Bm0E26tJ3gHYMgtrferzxHCn4v2oPkUKI) result: ![image](https://private-user-images.githubusercontent.com/12033811/380107356-c0517e1e-1ab1-498b-9d43-7229270126d1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5MzUyMjcsIm5iZiI6MTcyOTkzNDkyNywicGF0aCI6Ii8xMjAzMzgxMS8zODAxMDczNTYtYzA1MTdlMWUtMWFiMS00OThiLTlkNDMtNzIyOTI3MDEyNmQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI2VDA5Mjg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyZWM4NWM3YzQ2YTI1MDk4ZDk4YmMwMGFiNjY3M2Q2OWFmMzk5N2E4MGIwMTI4YmUwY2E5MjY4YTk5ODAxOTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tX_5uO4GdUdc-M4mx4th_6WS3fI_HK849HpUODwZYuQ)
> 
> ### Environment
> _No response_


The third picture shows that loss can be printed

**Comment by Edenzzzz** [comment]
Seems that your code is not the newest version. Could you pull the newest main branch and try again?

**Comment by liangzz1991** [comment]
> Seems that your code is not the newest version. Could you pull the newest main branch and try again?

torch 2.4.0 + colossalai 0.4.5 : same error

![image](https://github.com/user-attachments/assets/dcfbb329-7d03-4391-8fb8-1d94a7a07a0e)

![image](https://github.com/user-attachments/assets/3c09aaf6-f809-4206-a92b-8c03d85e80db)

**Comment by liangzz1991** [comment]
This is a simple piece of code that will give the same error：
`
import torch
import torch.nn as nn
from transformers import GPT2LMHeadModel,GPT2Config
from transformers import Qwen2VLForConditionalGeneration, AutoTokenizer, AutoProcessor

from colossalai.nn.optimizer import HybridAdam
import colossalai
from colossalai.booster import Booster
from colossalai.lazy import LazyInitContext
from colossalai.booster.plugin import GeminiPlugin
import os
# os.environ["RANK"] = "0"
# os.environ["LOCAL_RANK"] = "0"

# class GPTLMModel(nn.Module):

#     def __init__(self,
#                  hidden_size=768,
#                  num_layers=12,
#                  num_attention_heads=12,
#                  max_seq_len=1024,
#                  vocab_size=50257,
#                  checkpoint=False):
#         super().__init__()
#         self.checkpoint = checkpoint
#         self.model = GPT2LMHeadModel(
#             GPT2Config(n_embd=hidden_size,
#                        n_layer=num_layers,
#                        n_head=num_attention_heads,
#                        n_positions=max_seq_len,
#                        n_ctx=max_seq_len,
#                        vocab_size=vocab_size))
#         if checkpoint:
#             self.model.gradient_checkpointing_enable()

#     def forward(self, input_ids, attention_mask):
#         return self.model(input_ids=input_ids, attention_mask=attention_mask, use_cache=not self.checkpoint)[0]

# def gpt2_medium(checkpoint=False):
#     return GPTLMModel(hidden_size=1024, num_layers=24, num_attention_heads=16, checkpoint=checkpoint)

class GPTLMLoss(nn.Module):

    def __init__(self):
        super().__init__()
        self.loss_fn = nn.CrossEntropyLoss()

    def forward(self, logits, labels):
        shift_logits = logits[..., :-1, :].contiguous()
        shift_labels = labels[..., 1:].contiguous()
        return self.loss_fn(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
    
def get_data(batch_size, seq_len, vocab_size):
    input_ids = torch.randint(0, vocab_size, (batch_size, seq_len), device=torch.cuda.current_device())
    attention_mask = torch.ones_like(input_ids)
    return input_ids, attention_mask

def main():
    # args = parse_args()
    BATCH_SIZE = 8
    SEQ_LEN = 1024
    VOCAB_SIZE = 50257
    NUM_STEPS = 10
    colossalai.launch_from_torch()
    
    # build GPT model
    with LazyInitContext(default_device=torch.device('cuda')):
      # model = gpt2_medium(checkpoint=True)
      model = Qwen2VLForConditionalGeneration.from_pretrained("/data/liuxiaoyu/liuxiaoyu/models/Qwen2-VL-2B-Instruct", device_map="auto"
                                                              )
    # build criterion
    criterion = GPTLMLoss()
    optimizer = HybridAdam(model.parameters(), lr=0.001)

    torch.manual_seed(123)
    


    # Gemini + ZeRO DP
    plugin = GeminiPlugin(max_norm=1.0, initial_scale=2**5)
    booster = Booster(plugin=plugin)
    model, optimizer, criterion, _, _ = booster.boost(model, optimizer, criterion)

    torch.cuda.synchronize()
    model.train()
    for n in range(NUM_STEPS):
        print(n)
        # we just use randomly generated data here
        input_ids, attn_mask = get_data(BATCH_SIZE, SEQ_LEN, VOCAB_SIZE)
        optimizer.zero_grad()
        outputs = model(input_ids, attn_mask)
        # print(outputs)
        loss = criterion(outputs.logits, input_ids)
        booster.backward(loss, optimizer)
        optimizer.step()

    torch.cuda.synchronize()
    
if __name__ =='__main__':
    main()
`

**Comment by liangzz1991** [comment]
This is a simple piece of code that will give the same error：


import torch
import torch.nn as nn
from transformers import GPT2LMHeadModel,GPT2Config
from transformers import Qwen2VLForConditionalGeneration, AutoTokenizer, AutoProcessor

from colossalai.nn.optimizer import HybridAdam
import colossalai
from colossalai.booster import Booster
from colossalai.lazy import LazyInitContext
from colossalai.booster.plugin import GeminiPlugin

class GPTLMLoss(nn.Module):

    def __init__(self):
        super().__init__()
        self.loss_fn = nn.CrossEntropyLoss()

    def forward(self, logits, labels):
        shift_logits = logits[..., :-1, :].contiguous()
        shift_labels = labels[..., 1:].contiguous()
        return self.loss_fn(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
    
def get_data(batch_size, seq_len, vocab_size):
    input_ids = torch.randint(0, vocab_size, (batch_size, seq_len), device=torch.cuda.current_device())
    attention_mask = torch.ones_like(input_ids)
    return input_ids, attention_mask

def main():
    # args = parse_args()
    BATCH_SIZE = 8
    SEQ_LEN = 1024
    VOCAB_SIZE = 50257
    NUM_STEPS = 10
    colossalai.launch_from_torch()
    
    # build GPT model
    with LazyInitContext(default_device=torch.device('cuda')):
      # model = gpt2_medium(checkpoint=True)
      model = Qwen2VLForConditionalGeneration.from_pretrained("/data/liuxiaoyu/liuxiaoyu/models/Qwen2-VL-2B-Instruct", device_map="auto"
                                                              )
    # build criterion
    criterion = GPTLMLoss()
    optimizer = HybridAdam(model.parameters(), lr=0.001)

    torch.manual_seed(123)
    


    # Gemini + ZeRO DP
    plugin = GeminiPlugin(max_norm=1.0, initial_scale=2**5)
    booster = Booster(plugin=plugin)
    model, optimizer, criterion, _, _ = booster.boost(model, optimizer, criterion)

    torch.cuda.synchronize()
    model.train()
    for n in range(NUM_STEPS):
        print(n)
        # we just use randomly generated data here
        input_ids, attn_mask = get_data(BATCH_SIZE, SEQ_LEN, VOCAB_SIZE)
        optimizer.zero_grad()
        outputs = model(input_ids, attn_mask)
        # print(outputs)
        loss = criterion(outputs.logits, input_ids)
        booster.backward(loss, optimizer)
        optimizer.step()

    torch.cuda.synchronize()
    
if __name__ =='__main__':
    main()

**Comment by liangzz1991** [comment]
This is a simple piece of code that will give the same error：

````
```
import torch
import torch.nn as nn
from transformers import GPT2LMHeadModel,GPT2Config
from transformers import Qwen2VLForConditionalGeneration, AutoTokenizer, AutoProcessor

from colossalai.nn.optimizer import HybridAdam
import colossalai
from colossalai.booster import Booster
from colossalai.lazy import LazyInitContext
from colossalai.booster.plugin import GeminiPlugin
import os
# os.environ["RANK"] = "0"
# os.environ["LOCAL_RANK"] = "0"

# class GPTLMModel(nn.Module):

#     def __init__(self,
#                  hidden_size=768,
#                  num_layers=12,
#                  num_attention_heads=12,
#                  max_seq_len=1024,
#                  vocab_size=50257,
#                  checkpoint=False):
#         super().__init__()
#         self.checkpoint = checkpoint
#         self.model = GPT2LMHeadModel(
#             GPT2Config(n_embd=hidden_size,
#                        n_layer=num_layers,
#                        n_head=num_attention_heads,
#                        n_positions=max_seq_len,
#                        n_ctx=max_seq_len,
#                        vocab_size=vocab_size))
#         if checkpoint:
#             self.model.gradient_checkpointing_enable()

#     def forward(self, input_ids, attention_mask):
#         return self.model(input_ids=input_ids, attention_mask=attention_mask, use_cache=not self.checkpoint)[0]

# def gpt2_medium(checkpoint=False):
#     return GPTLMModel(hidden_size=1024, num_layers=24, num_attention_heads=16, checkpoint=checkpoint)

class GPTLMLoss(nn.Module):

    def __init__(self):
        super().__init__()
        self.loss_fn = nn.CrossEntropyLoss()

    def forward(self, logits, labels):
        shift_logits = logits[..., :-1, :].contiguous()
        shift_labels = labels[..., 1:].contiguous()
        return self.loss_fn(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
    
def get_data(batch_size, seq_len, vocab_size):
    input_ids = torch.randint(0, vocab_size, (batch_size, seq_len), device=torch.cuda.current_device())
    attention_mask = torch.ones_like(input_ids)
    return input_ids, attention_mask

def main():
    # args = parse_args()
    BATCH_SIZE = 8
    SEQ_LEN = 1024
    VOCAB_SIZE = 50257
    NUM_STEPS = 10
    colossalai.launch_from_torch()
    
    # build GPT model
    with LazyInitContext(default_device=torch.device('cuda')):
      # model = gpt2_medium(checkpoint=True)
      model = Qwen2VLForConditionalGeneration.from_pretrained("/data/liuxiaoyu/liuxiaoyu/models/Qwen2-VL-2B-Instruct", device_map="auto"
                                                              )
    # build criterion
    criterion = GPTLMLoss()
    optimizer = HybridAdam(model.parameters(), lr=0.001)

    torch.manual_seed(123)
    


    # Gemini + ZeRO DP
    plugin = GeminiPlugin(max_norm=1.0, initial_scale=2**5)
    booster = Booster(plugin=plugin)
    model, optimizer, criterion, _, _ = booster.boost(model, optimizer, criterion)

    torch.cuda.synchronize()
    model.train()
    for n in range(NUM_STEPS):
        print(n)
        # we just use randomly generated data here
        input_ids, attn_mask = get_data(BATCH_SIZE, SEQ_LEN, VOCAB_SIZE)
        optimizer.zero_grad()
        outputs = model(input_ids, attn_mask)
        # print(outputs)
        loss = criterion(outputs.logits, input_ids)
        booster.backward(loss, optimizer)
        optimizer.step()

    torch.cuda.synchronize()
    
if __name__ =='__main__':
    main()
```
````

**Comment by liangzz1991** [comment]
@wangbluo @gothicx

**Comment by Edenzzzz** [comment]
@botbw Any insights? Thanks

**Comment by liangzz1991** [comment]
> @botbw Any insights? Thanks
It was my mistake. The “Qwen2VLForConditionalGeneration” model input requires complete input of "input_ids", "attention_mask", "pixel_values", and "image_grid_thw". Although it can run only with "input_ids" and "attention_mask", the above will appear. The error I mentioned
